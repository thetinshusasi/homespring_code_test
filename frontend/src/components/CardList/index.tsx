import React from 'react'
import List from '@material-ui/core/List';
import Rating from '@material-ui/lab/Rating';


import useStyles from "./style"
import { Typography, Divider } from '@material-ui/core';
import CardItem from '../CardItem';
import Book from '../../models/books/Book';
import literals from '../../utils/literals';

export interface ICardList {
    books: Book[] | undefined
}
export default function CardList({ books }: ICardList) {
    const classes = useStyles();
    const content = books && books.length ? books?.map((book, index) => {
        return <span key={book.id + index}>
            <CardItem book={book} />
            <Divider variant="inset" component="li" />

        </span>
    }) : <Typography variant="h6" gutterBottom align='center'>
        {literals.noBooksMessage}
    </Typography>

    return (
        <>
            <List className={classes.root}>
                {content}
            </List >



        </>
    )

}
