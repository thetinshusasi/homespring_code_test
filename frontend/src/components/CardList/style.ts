import { createStyles, makeStyles, Theme } from "@material-ui/core";

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            width: '100%',
            backgroundColor: theme.palette.background.paper,
        },
        inline: {
            display: 'inline',
        },
    }),
);
export default useStyles