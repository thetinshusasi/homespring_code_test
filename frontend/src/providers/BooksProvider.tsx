import React, { useEffect, useState } from "react";
import useGetBooksBySearchTerm from "../hooks/useGetBooksBySearchTerm";
import Book from "../models/books/Book"

export interface BooksContextProps {
    books: Book[] | undefined,
    startIndex: number,
    kind: string,
    totalItems: number,
    error: any,
    loading: boolean,
    getBooksBySearchTerm: (val: string) => void
}
interface BooksProviderProps {
    children: React.ReactNode;
}
export const BooksContext = React.createContext<BooksContextProps>({
    books: undefined,
    error: false,
    getBooksBySearchTerm: () => { },
    kind: "",
    loading: false,
    startIndex: 0,
    totalItems: 0

});

const BooksProvider = ({ children }: BooksProviderProps) => {
    const { data, error, getBooksBySearchTerm, loading } = useGetBooksBySearchTerm();
    const { items: books = undefined, startIndex = 0, kind = "", totalItems = 0 } = data || {}

    const [value, setValue] = useState({
        books,
        error,
        getBooksBySearchTerm,
        loading,
        startIndex,
        totalItems,
        kind
    })

    useEffect(() => {
        const { items: books = undefined, startIndex = 0, kind = "", totalItems = 0 } = data || {}
        setValue({
            books,
            error,
            loading,
            startIndex,
            totalItems,
            kind,
            getBooksBySearchTerm

        })


    }, [data, loading, error])


    return (
        <BooksContext.Provider
            value={value}
        >
            {children}
        </BooksContext.Provider>
    );
};

export default BooksProvider;
