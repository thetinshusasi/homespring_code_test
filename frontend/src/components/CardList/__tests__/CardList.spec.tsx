/* eslint-disable testing-library/prefer-screen-queries */
/* eslint-disable jest/valid-expect */

import React from 'react';
import CardList, { ICardList } from '../index';
import BooksProvider, { BooksContextProps } from '../../../providers/BooksProvider';
import { render } from '@testing-library/react';

// has to be defined like this cannot be global

describe('CardList Component', () => {
    let props: ICardList;

    beforeEach(() => {
        props = {
            books: [{
                id: "1",
                volumeInfo: {
                    authors: ["a1"],
                    averageRating: 5,
                    categories: ['c1'],
                    description: "desc",
                    imageLinks: {
                        smallThumbnail: "smallImage",
                        thumbnail: 'image'
                    },
                    pageCount: 10,
                    publishedDate: "2020",
                    publisher: "publisher",
                    title: "title"

                }
            }]
        };
    });

    it('>> renders component', () => {

        const contextValue: BooksContextProps = {
            books: undefined,
            error: false,
            getBooksBySearchTerm: () => { },
            kind: "",
            loading: false,
            startIndex: 0,
            totalItems: 0
        };
        const { container } = render(
            <BooksProvider value={contextValue}>
                <CardList {...props} />
            </BooksProvider>
        );
        expect(container).toMatchSnapshot();
    });

    it('>> renders component and contains all props', async () => {
        const mockFn = jest.fn()
        const contextValue: BooksContextProps = {
            books: undefined,
            error: false,
            getBooksBySearchTerm: mockFn,
            kind: "",
            loading: false,
            startIndex: 0,
            totalItems: 0
        };
        const { container, getByText } = render(
            <BooksProvider value={contextValue}>
                <CardList {...props} />
            </BooksProvider>
        );
        expect(getByText('#10'))
        expect(getByText('2020'))
        expect(getByText('publisher'))
        expect(getByText('a1'))
        expect(getByText('c1'))
        expect(getByText('title'))
        expect(getByText('desc'))





    });


});
