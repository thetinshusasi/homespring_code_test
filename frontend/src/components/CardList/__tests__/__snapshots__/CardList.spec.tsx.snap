// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardList Component >> renders component 1`] = `
<div>
  <ul
    class="MuiList-root makeStyles-root-1 MuiList-padding"
  >
    <span>
      <li
        class="MuiListItem-root makeStyles-listItem-5 MuiListItem-gutters"
      >
        <div
          class="makeStyles-avatarContainer-3"
        >
          <div
            class="MuiAvatar-root MuiAvatar-square makeStyles-large-4"
          >
            <img
              class="MuiAvatar-img"
              src="smallImage"
            />
          </div>
          <div>
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-gutterBottom"
            >
              a1
            </h6>
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-gutterBottom"
            >
              c1
            </h6>
          </div>
        </div>
        <div
          class="makeStyles-listBody-6"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom"
          >
            title
          </h6>
          <span
            aria-label="5 Stars"
            class="MuiRating-root MuiRating-sizeLarge MuiRating-readOnly"
            role="img"
          >
            <span>
              <span
                class="MuiRating-icon MuiRating-iconFilled"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                  />
                </svg>
              </span>
            </span>
            <span>
              <span
                class="MuiRating-icon MuiRating-iconFilled"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                  />
                </svg>
              </span>
            </span>
            <span>
              <span
                class="MuiRating-icon MuiRating-iconFilled"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                  />
                </svg>
              </span>
            </span>
            <span>
              <span
                class="MuiRating-icon MuiRating-iconFilled"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                  />
                </svg>
              </span>
            </span>
            <span>
              <span
                class="MuiRating-icon MuiRating-iconFilled"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                  />
                </svg>
              </span>
            </span>
          </span>
          <h6
            class="MuiTypography-root makeStyles-subtitle-9 MuiTypography-subtitle2 MuiTypography-gutterBottom"
          >
            <span>
              #10
            </span>
            <span>
              2020 
            </span>
            <span>
              publisher
            </span>
                           
          </h6>
          <h6
            class="MuiTypography-root makeStyles-maxLines-8 MuiTypography-subtitle1 MuiTypography-gutterBottom"
          >
            desc
          </h6>
        </div>
      </li>
      <li
        class="MuiDivider-root MuiDivider-inset"
        role="separator"
      />
    </span>
  </ul>
</div>
`;
